keps/sig-release/20190118-creating-packages.md

```diff
@@ -0,0 +1,417 @@
+
+- The tooling must support signing the generated packages
+
+- The documentation for the process of building packages, local one-off builds
+  and as part of the release process, must be excellent.
+
+- Any other option to build packages must be retired eventually
+
+  There should be documentation how and where the old packaging option can be
+  found and how they can be used for cases where someone wants to build
+  packages before the new tools were introduced.
+
+- It must be equally easy to build packages from
+  - the currently checked out revision (any revision after the tooling has been
+    introduced into `k/k`)
+  - a pre-build binary (pulling from some well-defined staging bucket or that like)
+
+- It must be containerized and must not depend on anything more then docker on
+  the host
+
+### Non-Goals
+
+- Uploading, publishing, ... of packages
+- Clean up [`k/k/build/`][k-k] (remove outdated scripts & documentation) other
+  then "competitor" packaging mechanism
+
+## Proposal
+
+All the following steps are only considered done when there is proper documentation on
+- how to use those tools
+- how they interact with each other
+- how to (re-)build them
+- how to update them
+- how their configuration works
+
+**Note**: This section refers to a couple of specifics that might be seen as
+implementation details in here (e.g. references to a specific tools or specific
+filesystem layouts). Those are just there to illustrate in more detail what we
+aim towards, the details and actual tools can absolutely change in the actual
+implementation. 
+
+- [ ] Define and document the package configuration
+
+  The data referred to in this KEP as package configuration might be something
+  like <a id="nfpm-config">a `nfpm` (compatible) configuration file</a> plus a set
+  of additional files referred to in the [main configuration
+  file](#nfpm-config), potentially organised in subdirectories
+
+  ```sh
+  ./main.conf
+  ./linux/50-kubeadm.conf
+  ./linux/amd64/50-kubeadm.conf
+  [...]
+  ```
+
+- [ ] Create a tool that can create **one** package configuration
+   - must be a container image, the Dockerfile must be committed to `k/k`
+   - must take the following inputs:
+     - the package configuration
+     - the pre-compiled binarie(s) 
+     - the version the binaries are compiled from
+   - might take as input:
+     - the platform(s) to package for (default: [all supported platforms](#plat-list))
+   - must create the following outputs:
+     - one package per platform (`linux/386`, `darwin/amd64`, ...) and package type (`rpm`, `deb`)
+
+   Example run:
+   ```bash
+   docker build -t k8s-pkg-builder build/pkgs
+   docker run \
+     -v ${PWD}/_output/dockerized/bin:/in \
+     -v ${PWD}/_output/pkgs/nightly:/out \
+     -v ${PWD}/build/pkgs/kubeadm:/conf \
+     -e KUBE_BUILD_VERSION='1.12.6~beta.0.1+12187918e930d4+dirty' \
+     -e KUBE_BUILD_PLATFORMS='linux/amd64 linux/386' \
+     k8s-pkg-builder
```

Comment:
minor: we should consider just using the kubernetes docker build image & machinery then, we already have a lot there including multi-arch tooling. shipping another docker image for this seems non-ideal.

Reply:
Oh that'd be great! However I don't have too much insights on how you manage images over there at sig-testing (that's what you are talking about, right?), which images you already have, ...
I will try to get an overview on that.